lib/active_merchant/billing/gateways/payflow.rb

```diff
@@ -202,6 +206,29 @@ def build_credit_card_request(action, money, credit_card, options)
         add_level_two_three_fields(xml.target!, options)
       end
 
+      def add_mpi_3ds(xml, three_d_secure_options)  
+        # structure as per https://developer.paypal.com/api/nvp-soap/payflow/3d-secure-mpi/
+        authentication_id = three_d_secure_options[:authentication_id]
+        authentication_status = three_d_secure_options[:authentication_response_status]
+
+        eci = three_d_secure_options[:eci]
+        cavv = three_d_secure_options[:cavv]
+        xid = three_d_secure_options[:xid]
+        version = three_d_secure_options[:version]
+
+        # 3DS2 only
+        ds_transaction_id = three_d_secure_options[:ds_transaction_id] if version_2_or_newer?(three_d_secure_options)
+
+        xml.tag! 'ExtData', 'Name' => 'AUTHENTICATION_ID', 'Value' => authentication_id unless authentication_id.blank?
+        xml.tag! 'ExtData', 'Name' => 'AUTHENTICATION_STATUS', 'Value' => authentication_status unless authentication_status.blank?
+
+        xml.tag! 'ExtData', 'Name' => 'CAVV', 'Value' => cavv unless cavv.blank?
+        xml.tag! 'ExtData', 'Name' => 'ECI', 'Value' => eci unless eci.blank?
+        xml.tag! 'ExtData', 'Name' => 'XID', 'Value' => xid unless xid.blank?
+        xml.tag! 'ExtData', 'Name' => 'THREEDSVERSON', 'Value' => version unless version.blank?
+        xml.tag! 'ExtData', 'Name' => 'DSTRANSACTIONID', 'Value' => ds_transaction_id unless ds_transaction_id.blank?
```

Comment:
Quick look at the docs shows that we send certain fields depending on "cardholder enrollment in a 3-D Secure program", should/can we add logic here for that or are we expecting that to be accounted for earlier in the process?

Reply:
yeah, I felt setting those values would depend on the client application, but will look into incorporating that logic here too.

Reply:
If it doesn't make sense to place here then let's not force it. I just wanted to double check!